x-pack/plugins/apm/public/components/shared/EnvironmentFilter/index.tsx

```diff
@@ -0,0 +1,123 @@
+
+const anyEnvironmentValue = 'All';
+const nullEnvironmentValue = 'Null';
+
+function updateEnvironmentUrl(
+  location: ReturnType<typeof useLocation>,
+  environment?: string
+) {
+  const nextEnvironmentQueryParam =
+    environment !== anyEnvironmentValue ? environment : undefined;
+  history.push({
+    ...location,
+    search: fromQuery({
+      ...toQuery(location.search),
+      environment: nextEnvironmentQueryParam
+    })
+  });
+}
+
+const OptionInputContainer = styled.div`
+  min-width: 128px;
+`;
+
+const EnvironmentFilterLabel = styled.div`
+  height: 100%;
+  border-width: 1px 0 1px 1px;
+  border-style: solid;
+  border-color: #e1e1e1;
+  background: #eee;
+  font-size: ${fontSizes.small};
+  padding: ${px(units.minus)};
+`;
+
+export const EnvironmentFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { uiFilters } = useUiFilters();
+  const { start, end } = urlParams;
+  const { environment } = uiFilters;
+  const { data = [], status = 'loading' } = useFetcher(
+    () => {
+      if (start && end) {
+        return loadServiceEnvironments({ start, end });
+      }
+    },
+    [start, end]
+  );
+
+  const availableEnvironments = [
+    {
+      value: anyEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.allLabel', {
+        defaultMessage: 'All'
+      })
+    },
+    ...data.map(environment => ({ value: environment, label: environment })),
+    {
+      value: nullEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.undefinedLabel', {
+        defaultMessage: 'Undefined'
+      })
+    }
+  ];
+
+  if (
+    environment &&
+    environment !== anyEnvironmentValue &&
+    environment !== nullEnvironmentValue &&
+    status === 'loading'
+  ) {
+    availableEnvironments.push({ value: environment, label: environment });
+  }
+
+  const [valueOfSelected, setValueOfSelected] = useState(
+    environment || anyEnvironmentValue
+  );
+
+  useEffect(
+    () => {
+      setValueOfSelected(environment || anyEnvironmentValue);
+    },
+    [environment]
+  );
+
+  return (
+    <EuiFlexGroup justifyContent="flexEnd" alignItems="stretch" gutterSize="none">
+      <EuiFlexItem>
+        <EnvironmentFilterLabel>
+          {i18n.translate('xpack.apm.filter.environment.label', {
+            defaultMessage: 'environment'
+          })}
+        </EnvironmentFilterLabel>
+      </EuiFlexItem>
+      <EuiFlexItem>
+        <EuiSuperSelect
+          options={availableEnvironments.map(({ value, label }) => ({
+            value,
+            inputDisplay: <OptionInputContainer>{label}</OptionInputContainer>,
+            dropdownDisplay: <>{label}</>
+          }))}
+          valueOfSelected={valueOfSelected}
+          onChange={(value: string) => {
+            updateEnvironmentUrl(location, value);
+          }}
+          isLoading={status === 'loading'}
+        />
+      </EuiFlexItem>
+    </EuiFlexGroup>
+  );
+};
```

Comment:
What do you think about this?

```tsx
export function ServiceEnvFilter() {
  const location = useLocation();
  const { urlParams } = useUrlParams();
  const { uiFilters } = useUiFilters();
  const { start, end } = urlParams;
  const { environment } = uiFilters;
  const { data: environments = [], status } = useFetcher(
    () => {
      if (start && end) {
        return loadServiceEnvironments({ start, end });
      }
    },
    [start, end]
  );

  const availableEnvironments = [
    {
      value: undefined,
      label: i18n.translate('xpack.apm.filter.environment.allLabel', {
        defaultMessage: 'All'
      })
    },
    ...environments.map(environment => {
      if (environment === ENVIRONMENT_NOT_AVAILABLE) {
        return {
          value: environment,
          label: i18n.translate(
            'xpack.apm.filter.environment.notAvailableLabel',
            {
              defaultMessage: 'N/A'
            }
          )
        };
      }

      return { value: environment, label: environment };
    })
  ];

  return (
    <EuiSuperSelect
      options={availableEnvironments.map(({ value, label }) => ({
        value,
        inputDisplay: <OptionInput>{label}</OptionInput>,
        dropdownDisplay: <>{label}</>
      }))}
      valueOfSelected={environment}
      onChange={environment => {
        history.push({
          ...location,
          search: fromQuery({
            ...toQuery(location.search),
            environment
          })
        });
      }}
      isLoading={status === 'loading'}
    />
  );
}
```

`ENVIRONMENT_NOT_AVAILABLE` will be a shared constant that the backend uses as the `missing` key: https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html#_missing_value_14